mixer/adapter/solarwinds/appoptics/client.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package appoptics
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"istio.io/istio/mixer/pkg/adapter"
+)
+
+// ServiceAccessor defines an interface for talking to  via domain-specific service constructs
+type ServiceAccessor interface {
+	// MeasurementsService implements an interface for dealing with  Measurements
+	MeasurementsService() MeasurementsCommunicator
+}
+
+const (
+	// MeasurementPostMaxBatchSize defines the max number of Measurements to send to the API at once
+	MeasurementPostMaxBatchSize = 1000
+	defaultBaseURL              = "https://api.appoptics.com/v1/"
+	defaultMediaType            = "application/json"
+)
+
+// Client implements ServiceAccessor
+type Client struct {
+	// baseURL is the base endpoint of the remote  service
+	baseURL *url.URL
+	// client is the http.Client singleton used for wire interaction
+	client *http.Client
+	// token is the private part of the API credential pair
+	token string
+	// measurementsService embeds the client and implements access to the Measurements API
+	measurementsService MeasurementsCommunicator
+
+	logger adapter.Logger
+}
+
+// ErrorResponse represents the response body returned when the API reports an error
+type ErrorResponse struct {
+	// Errors holds the error information from the API
+	Errors interface{} `json:"errors"`
+}
+
+// NewClient creates a new client to interact with appoptics
+func NewClient(token string, logger adapter.Logger) *Client {
+	baseURL, _ := url.Parse(defaultBaseURL)
+	c := &Client{
+		client:  new(http.Client),
+		token:   token,
+		baseURL: baseURL,
+		logger:  logger,
+	}
+	c.measurementsService = &MeasurementsService{c, logger}
+
+	c.client.Transport = &http.Transport{
+		MaxIdleConnsPerHost: 10,
+	}
+
+	return c
+}
+
+// NewRequest standardizes the request being sent
+func (c *Client) NewRequest(method, path string, body interface{}) (*http.Request, error) {
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	requestURL := c.baseURL.ResolveReference(rel)
+
+	var buffer io.ReadWriter
+
+	if body != nil {
+		buffer = &bytes.Buffer{}
+		encodeErr := json.NewEncoder(buffer).Encode(body)
+		if encodeErr != nil {
+			dumpMeasurements(body, c.logger)
+			return nil, encodeErr
+		}
+
+	}
+	req, err := http.NewRequest(method, requestURL.String(), buffer)
+
+	if err != nil {
+		return nil, err
```

Comment:
worth adding context to the error here?  maybe `return nil, fmt.Errorf("failed connecting to appoptics: %v", err)` ?

Reply:
sure